From dcec8dcf6bff7b6cdd5711f7c409f5a26f8e150e Mon Sep 17 00:00:00 2001
From: Ravindra Reddy K <ravindra@marvell.com>
Date: Mon, 27 Nov 2017 14:22:09 +0530
Subject: [PATCH 2123/2241] fix: net: mv88e6xxx: set cpu, mirror, ingress and
 egress monitor destination port for peridot.

Fixed bug related to setting CPU destination, Mirror destination,
ingress monitor destination and egress monitor destination ports.
Unlike other soho devices where it is set by direct register write of
Global 1 register 0x1A, in peridot it is set indirectly using a pointer
to the type of destination port in Global 1 register 0x1A.

Change-Id: I5ce4f58d54a50f6af945a704016d77cd82fc2c94
Signed-off-by: Ravindra Reddy K <ravindra@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/46888
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/47450
---
 drivers/net/dsa/mv88e6xxx.c | 43 +++++++++++++++++++++++++++++++++++++------
 drivers/net/dsa/mv88e6xxx.h |  8 +++++++-
 2 files changed, 44 insertions(+), 7 deletions(-)

diff --git a/drivers/net/dsa/mv88e6xxx.c b/drivers/net/dsa/mv88e6xxx.c
index eef6827..76f8feb 100644
--- a/drivers/net/dsa/mv88e6xxx.c
+++ b/drivers/net/dsa/mv88e6xxx.c
@@ -2985,13 +2985,44 @@ static int mv88e6xxx_setup_global(struct mv88e6xxx_priv_state *ps)
 
 	/* Configure the upstream port, and configure it as the port to which
 	 * ingress and egress and ARP monitor frames are to be sent.
+	 *
+	 * For 6390 family, cpu destination port, mirror destination, ingress and egress
+	 * destination ports are set indirectly using a pointer to respctive type of destination.
+	 * For other family of devices, it is set by directly writing to the register.
 	 */
-	reg = upstream_port << GLOBAL_MONITOR_CONTROL_INGRESS_SHIFT |
-		upstream_port << GLOBAL_MONITOR_CONTROL_EGRESS_SHIFT |
-		upstream_port << GLOBAL_MONITOR_CONTROL_ARP_SHIFT;
-	err = _mv88e6xxx_reg_write(ps, REG_GLOBAL, GLOBAL_MONITOR_CONTROL, reg);
-	if (err)
-		return err;
+	if (mv88e6xxx_6390_family(ps)) {
+		/* Configure CPU destination port */
+		reg = GLOBAL_MMC_UPDATE | (GLOBAL_MMC_CPUDEST_PTR << GLOBAL_MMC_PTR_SHIFT) | upstream_port;
+		err = _mv88e6xxx_reg_write(ps, REG_GLOBAL, GLOBAL_MONITOR_CONTROL, reg);
+		if (err)
+			return err;
+
+		/* Configure Mirror destination port */
+		reg = GLOBAL_MMC_UPDATE | (GLOBAL_MMC_MIRRORDEST_PTR << GLOBAL_MMC_PTR_SHIFT) | upstream_port;
+		err = _mv88e6xxx_reg_write(ps, REG_GLOBAL, GLOBAL_MONITOR_CONTROL, reg);
+		if (err)
+			return err;
+
+		/* Configure Ingress Monitor destination port */
+		reg = GLOBAL_MMC_UPDATE | (GLOBAL_MMC_INGRESSDEST_PTR << GLOBAL_MMC_PTR_SHIFT) | upstream_port;
+		err = _mv88e6xxx_reg_write(ps, REG_GLOBAL, GLOBAL_MONITOR_CONTROL, reg);
+		if (err)
+			return err;
+
+		/* Configure Egress Monitor destination port */
+		reg = GLOBAL_MMC_UPDATE | (GLOBAL_MMC_EGRESSDEST_PTR << GLOBAL_MMC_PTR_SHIFT) | upstream_port;
+		err = _mv88e6xxx_reg_write(ps, REG_GLOBAL, GLOBAL_MONITOR_CONTROL, reg);
+		if (err)
+			return err;
+
+	} else {
+		reg = upstream_port << GLOBAL_MONITOR_CONTROL_INGRESS_SHIFT |
+			upstream_port << GLOBAL_MONITOR_CONTROL_EGRESS_SHIFT |
+			upstream_port << GLOBAL_MONITOR_CONTROL_ARP_SHIFT;
+		err = _mv88e6xxx_reg_write(ps, REG_GLOBAL, GLOBAL_MONITOR_CONTROL, reg);
+		if (err)
+			return err;
+	}
 
 	/* Disable remote management, and set the switch's DSA device number. */
 	err = _mv88e6xxx_reg_write(ps, REG_GLOBAL, GLOBAL_CONTROL_2,
diff --git a/drivers/net/dsa/mv88e6xxx.h b/drivers/net/dsa/mv88e6xxx.h
index 9addb3c..df060e8 100644
--- a/drivers/net/dsa/mv88e6xxx.h
+++ b/drivers/net/dsa/mv88e6xxx.h
@@ -263,7 +263,13 @@
 #define GLOBAL_MONITOR_CONTROL_ARP_SHIFT	4
 #define GLOBAL_MONITOR_CONTROL_MIRROR_SHIFT	0
 #define GLOBAL_MONITOR_CONTROL_ARP_DISABLED	(0xf0)
-#define GLOBAL_CONTROL_2	0x1c
+#define GLOBAL_MMC_UPDATE			BIT(15)	/* 6390 */
+#define GLOBAL_MMC_CPUDEST_PTR			0x30	/* 6390 */
+#define GLOBAL_MMC_MIRRORDEST_PTR		0x22	/* 6390 */
+#define GLOBAL_MMC_EGRESSDEST_PTR		0x21	/* 6390 */
+#define GLOBAL_MMC_INGRESSDEST_PTR		0x20	/* 6390 */
+#define GLOBAL_MMC_PTR_SHIFT			8	/* 6390 */
+#define GLOBAL_CONTROL_2			0x1c
 #define GLOBAL_CONTROL_2_NO_CASCADE		0xe000
 #define GLOBAL_CONTROL_2_MULTIPLE_CASCADE	0xf000
 
-- 
2.7.4

